tests/pipeline/test_execution_stage.py

```diff
@@ -0,0 +1,190 @@
+    QuotaAllowance,
+)
+from snuba.query.data_source.simple import Entity, Table
+from snuba.query.dsl import and_cond, column, equals, literal
+from snuba.query.expressions import Column, FunctionCall
+from snuba.query.logical import Query as LogicalQuery
+from snuba.query.query_settings import HTTPQuerySettings
+from snuba.querylog.query_metadata import SnubaQueryMetadata
+from snuba.request import Request
+from snuba.utils.metrics.timer import Timer
+from snuba.utils.schemas import UUID, String, UInt
+
+
+class MockAllocationPolicy(AllocationPolicy):
+    def _additional_config_definitions(self) -> list[AllocationPolicyConfig]:
+        # Define policy specific config definitions, these will be used along
+        # with the default definitions of the base class. (is_enforced, is_active)
+        return []
+
+    def _get_quota_allowance(
+        self, tenant_ids: dict[str, str | int], query_id: str
+    ) -> QuotaAllowance:
+        return QuotaAllowance(can_run=True, max_threads=1, explanation={})
+
+    def _update_quota_balance(
+        self,
+        tenant_ids: dict[str, str | int],
+        query_id: str,
+        result_or_error: QueryResultOrError,
+    ) -> None:
+        self.did_update = True
+
+
+policy = MockAllocationPolicy(
+    StorageKey("mystorage"),
+    required_tenant_types=["organization_id", "referrer"],
+    default_config_overrides={},
+)
+
+mockmetadata = SnubaQueryMetadata(
+    Request(
+        "",
+        {},
+        LogicalQuery(
+            from_clause=Entity(key=EntityKey.TRANSACTIONS, schema=EntityColumnSet([]))
+        ),
+        HTTPQuerySettings(),
+        AttributionInfo(
+            get_app_id("blah"), {"tenant_type": "tenant_id"}, "blah", None, None, None
+        ),
+        "",
+    ),
+    "blah",
+    Timer("woof"),
+)
+
+ch_query = Query(
+    from_clause=Table(
+        "transactions_local",
+        ColumnSet(
+            [
+                ("event_id", UUID()),
+                ("project_id", UInt(64)),
+                ("transaction_name", String()),
+                ("duration", UInt(32)),
+            ]
+        ),
+        storage_key=StorageKey.TRANSACTIONS,
+        allocation_policies=[policy],
+    ),
+    selected_columns=[
+        SelectedExpression(
+            "avg(duration)",
+            FunctionCall(
+                "_snuba_avg(duration)",
+                "avg",
+                (Column("_snuba_duration", None, "duration"),),
+            ),
+        )
+    ],
+    condition=and_cond(
+        equals(column("transaction_name"), literal("dog")),
+        equals(column("project_id"), literal(1)),
+    ),
+    limit=1000,
+)
+
+
+@pytest.mark.clickhouse_db
+@pytest.mark.redis_db
+def test_basic() -> None:
+    attinfo = AttributionInfo(
+        get_app_id("blah"), {"tenant_type": "tenant_id"}, "blah", None, None, None
+    )
+    settings = HTTPQuerySettings()
+    timer = Timer("test")
+    res = ExecutionStage(attinfo, query_metadata=mockmetadata).execute(
```

Comment:
this query_metadata object gets updated on execution, the variable local to the test and you should probably check what's in it after this is done